motionplan/rrtStarConnect.go

```diff
@@ -176,64 +212,111 @@ func (mp *rrtStarConnectMotionPlanner) rrtBackgroundRunner(ctx context.Context,
-	tree rrtMap,
-	target []referenceframe.Input,
+	ctx context.Context,
+	rrtMap map[node]node,
+	target node,
 	mchan chan node,
 ) {
-	if validTarget := mp.checkInputs(target); !validTarget {
-		mchan <- nil
-		return
-	}
-	// iterate over the k nearest neighbors and find the minimum cost to connect the target node to the tree
-	neighbors := kNearestNeighbors(mp.planOpts, tree, &basicNode{q: target}, mp.algOpts.NeighborhoodSize)
-	minCost := math.Inf(1)
-	minIndex := -1
-	for i, thisNeighbor := range neighbors {
-		cost := thisNeighbor.node.Cost() + thisNeighbor.dist
-		if mp.checkPath(thisNeighbor.node.Q(), target) {
-			minIndex = i
-			minCost = cost
-			// Neighbors are returned ordered by their costs. The first valid one we find is best, so break here.
-			break
+	// This should iterate until one of the following conditions:
+	// 1) we have reached the target
+	// 2) the request is cancelled/times out
+	// 3) we are no longer approaching the target and our "best" node is further away than the previous best
+	// 4) further iterations change our best node by close-to-zero amounts
+	// 5) we have iterated more than maxExtendIter times
+	near := kNearestNeighbors(mp.planOpts, rrtMap, &basicNode{q: target.Q()}, mp.algOpts.NeighborhoodSize)[0].node
+	oldNear := near
+	for i := 0; i < maxExtendIter; i++ {
+		select {
+		case <-ctx.Done():
+			mchan <- oldNear
+			return
+		default:
 		}
-	}
 
-	// add new node to tree as a child of the minimum cost neighbor node if it was reachable
-	if minIndex == -1 {
-		mchan <- nil
-		return
-	}
-	targetNode := &basicNode{q: target, cost: minCost}
-	tree[targetNode] = neighbors[minIndex].node
-
-	// rewire the tree
-	for i, thisNeighbor := range neighbors {
-		// dont need to try to rewire minIndex, so skip it
-		if i == minIndex {
-			continue
+		dist := mp.planOpts.DistanceFunc(&Segment{StartConfiguration: near.Q(), EndConfiguration: target.Q()})
+		oldDist := mp.planOpts.DistanceFunc(&Segment{StartConfiguration: oldNear.Q(), EndConfiguration: target.Q()})
+		switch {
+		case dist < mp.algOpts.JointSolveDist:
+			mchan <- near
+			return
+		case dist > oldDist:
+			mchan <- oldNear
+			return
 		}
 
-		// check to see if a shortcut is possible, and rewire the node if it is
-		connectionCost := mp.planOpts.DistanceFunc(&Segment{
-			StartConfiguration: thisNeighbor.node.Q(),
-			EndConfiguration:   targetNode.Q(),
-		})
-		cost := connectionCost + targetNode.Cost()
-
-		// If 1) we have a lower cost, and 2) the putative updated path is valid
-		if cost < thisNeighbor.node.Cost() && mp.checkPath(target, thisNeighbor.node.Q()) {
-			// Alter the cost of the node
-			// This needs to edit the existing node, rather than make a new one, as there are pointers in the tree
-			thisNeighbor.node.SetCost(cost)
-			tree[thisNeighbor.node] = targetNode
+		oldNear = near
+		newNear := make([]referenceframe.Input, 0, len(near.Q()))
+
+		// alter near to be closer to target
+		for j, nearInput := range near.Q() {
```

Comment:
This is all to interpolate?  Shouldn't there be a function we should be able to call that does this for us?